src/Microsoft.AspNetCore.Sockets.Client.Http/WebSocketsTransport.cs

```diff
@@ -77,43 +78,84 @@ public WebSocketsTransport(HttpOptions httpOptions, ILoggerFactory loggerFactory
             _logger.StartTransport(Mode.Value);
 
             await Connect(url);
-            var sendTask = SendMessages();
-            var receiveTask = ReceiveMessages();
 
             // TODO: Handle TCP connection errors
             // https://github.com/SignalR/SignalR/blob/1fba14fa3437e24c204dfaf8a18db3fce8acad3c/src/Microsoft.AspNet.SignalR.Core/Owin/WebSockets/WebSocketHandler.cs#L248-L251
-            Running = Task.WhenAll(sendTask, receiveTask).ContinueWith(t =>
+            Running = ProcessSocketAsync(_webSocket);
+        }
+
+        private async Task ProcessSocketAsync(WebSocket socket)
+        {
+            using (socket)
             {
-                _webSocket.Dispose();
-                _logger.TransportStopped(t.Exception?.InnerException);
+                // Begin sending and receiving. Receiving must be started first because ExecuteAsync enables SendAsync.
+                var receiving = StartReceiving(socket);
+                var sending = StartSending(socket);
 
-                _application.Output.Complete(t.Exception?.InnerException);
-                _application.Input.Complete();
+                // Wait for send or receive to complete
+                var trigger = await Task.WhenAny(receiving, sending);
+
+                if (trigger == receiving)
+                {
+                    // We're waiting for the application to finish and there are 2 things it could be doing
+                    // 1. Waiting for application data
+                    // 2. Waiting for a websocket send to complete
+
+                    // Cancel the application so that ReadAsync yields
+                    _application.Input.CancelPendingRead();
+
+                    using (var delayCts = new CancellationTokenSource())
+                    {
+                        var resultTask = await Task.WhenAny(sending, Task.Delay(_closeTimeout, delayCts.Token));
```

Comment:
Could probably just unconditionally cancel the CTS here, since if it's the one that fired, the cancellation won't mean anything anyway. But I don't really mind either way.